ouroboros-consensus/docs/HardFork.md

```diff
@@ -0,0 +1,211 @@
+candidate proposal
+   --> (2k slots later)
+stably-candidated proposal    (Frisby: stably-nominated?)
+   --> (epoch transition)
+adopted proposal
+```
+
+### Initiating the hard fork
+
+Proposals to initiate the hard fork can be submitted and voted on before all
+core nodes are ready. After all, once a proposal is "stably-confirmed", it will
+effectively remain so indefinitely until nodes endorse it (or it gets superseded
+by another proposal). This means that nodes can vote to initiate the hard fork,
+_then_ wait for everybody to update their software, and once updated, the
+proposal is endorsed and eventually the hard fork is initiated.
+
+Endorsement is somewhat implicit. The node operator does not submit an explicit
+"endorsement transaction", but instead restarts the node (probably after a
+software update that makes the node ready to support the hard fork) with a new
+protocol version (as part of a config file or command line parameter), which
+then gets included in the blocks that the node produces (this value is part of
+the static `ByronConfig`: `byronProtocolVersion`).
+
+(Note that a node restart is necessary for _any_ change to a protocol parameter,
+even though most parameters do not require any change to the software at all.)
+
+### Software version
+
+The Byron header also records a software version (`headerSoftwareVersion`). This
+is a legacy concern only, and is present in but ignored by the current Byron
+implementation, and entirely absent from the Byron specification.
+
+## Shelley
+
+### Moment of the hard fork
+
+Similar to the Byron ledger, the Shelley ledger provides a "current protocol
+version", but it is a two-tuple (not a three-tuple), containing only a
+`hard fork` component and `soft fork` component:
+
+```haskell
+_protocolVersion :: (Natural, Natural)
+```
+
+in `PParams` (currently, module `PParams` in
+`chain-and-ledger/executable-spec/src/PParams.hs`).
+
+The hard fork from Shelley to its successor (Goguen?) will be initiated
+once the hard fork component of this version gets incremented.
+
+### The update mechanism for the protocol version
+
+The update mechanism in Shelley is simpler than it is in Byron. There is no
+distinction between votes and proposals: to "vote" for a proposal one merely
+submits the exact same proposal. There is also no separate endorsement step.
+The procedure is as follows:
+
+1. Proposals (change parameter `x` to value `y` are submitted to the chain)
+2. "Voting" (submitting of proposals) ends `2k` slots before the end of the
+   epoch (that `2k` might need to be adjusted pending more detailed analysis
+   from the researchers -- the spec records this as a single specification
+   parameter (not protocol parameter!) called `slotsPrior`).
+3. By the end of the epoch, if a majority (for some definition of majority) is
+   reached for _all_ proposed parameters, then _all_ parameters take their new
+   value at the start of the next epoch. If no majority is reached on _any one_
```

Comment:
> If no majority is reached on _any one_ of the parameters

This is an aside. But I suspect the wording here is slightly inaccurate or else that we have an additional validation rule that prevents contemporary proposals from altering the same parameter. Specifically: if two proposals adjust the same parameter and one of them is confirmed but the other is not, then that unconfirmed proposal would still spoil all the proposals, right?

Reply:
I think it is correct as written. "proposals" and "votes" are really the same thing in Shelley. Each core node is allowed to include a update proposal during the voting period (any time before "2k" slots from the end of the epoch). either there are at least five proposals that agreed exactly, or there is not.

Reply:
in particular, if four nodes want `{foo = 1, bar = 2}` and three nodes `{foo = 1}`, there is _no consensus_ and nothing is changed.

Reply:
@JaredCorduan It sounds like you're saying that, in Shelley:

  * Each node is allowed per epoch to propose zero or one udpates to the parameter values, not two or more proposals.

  * At an epoch transition, an update is only adopted if every node had proposed that same exact update "early enough" before the transition.

Is that right? Disclaimer: I haven't yet looked at the Shelley spec; I've only been testing Byron as of yet, my brain's bandwith, and so on.

Reply:
> Each node is allowed per epoch to propose zero or one udpates to the parameter values, not two or more proposals.

Not sure which distinction you're drawing here?

Reply:
Disclaimer: This is a pretty low-level concern, but I think it's a reasonable interpretation of the current text.

Consider this scenario. There are 3 nodes.

  * Node A proposes `foo = 1`.
  * Node B proposes `foo = 1`.
  * Node C proposes `foo = 2`.

If 2/3 is a sufficient majority, then the text as written suggests that `foo = 1` will be adopted. The text currently says "If no majority is reached on any one of the parameters". Even though there is a failed proposal for `foo`, there is also a confirmed proposal for `foo`; so "every one of the parameters" that has reached a majority, but not "every one of the proposals" has.

If I correctly interpreted Jared's comment and unanimity is required, then the same sort of scenario would have to look like this.

  * Node A proposes `foo = 1`.
  * Node A also proposes `foo = 2`.
  * Node B proposes `foo = 1`.
  * Node C proposes `foo = 1`.

But if I also correctly interpreted his comment to mean both unanimity is required and also that the same node can't put out more than one proposal in a given epoch, then this kind of scenario is impossible and the current wording in the text is ultimately fine.

Reply:
@nfrisby - a majority is required, not unanimity. each core node can provide multiple updates during each epoch, but the last one overrides any previous ones. So in your last example, `foo = 1` is in fact adopted, since it gets 2/3 votes.

Reply:
Aha! I was wrong on both accounts :). I think we should revise this paragraph. Edsko and I didn't have a correct understanding of Shelley protocol parameter updates when we wrote it.

I've now skimmed the Shelley spec for this purpose. But the pdf notation puns on the exact types pretty heavily for the sake of brevity, so I've turned to the code to find the details. As of commit e09a3efa9f428de474384856c0311088b1503819, I see the following behavior and code snippets.

[There is a tl;dr at the bottom. This comment got long.]

The `CHAIN` rule starts with an empty `PPUpdate Map.empty` as a field in the `UTxOState`.

```
newtype PPUpdate crypto = PPUpdate (Map (GenKeyHash crypto) PParamsUpdate)

newtype PParamsUpdate = PParamsUpdate { ppmSet :: Set Ppm }
  deriving (Show, Eq, NoUnexpectedThunks)

data Ppm = MinFeeA Integer
  | MinFeeB Natural
  | MaxBBSize Natural
  | ...
  deriving (Show, Ord, Eq, Generic)
```

So, at all times, we have a function isomorphic to `GenKeyHash crypto -> Maybe
(Set Ppm)`.

I don't see anywhere in `shelley/chain-and-ledger/executable-spec/src` that indicates how a node would update its `Set Ppm`, how it would "emit" a `Ppm`. Is there a separate repository that has code to do that? Or is that not yet implemented? I only see the `PPUP` rule, but that takes a whole chain-level `PPUpdate` as its signal, whereas I'm trying to find what it looks like from the perspective of a lone node (operator).

In the `EPOCH` rule, we count up to see how many genesis keys have the same set of `Ppm`s.

```
votedValuePParams
  :: PPUpdate crypto
  -> PParams
  -> Int
  -> Maybe PParams
votedValuePParams (PPUpdate ppup) pps quorumN =
  let
    incrTally vote tally = 1 + Map.findWithDefault 0 vote tally
    votes = Map.foldr
              (\vote tally -> Map.insert vote (incrTally vote tally) tally)
              (Map.empty :: Map PParamsUpdate Int)
              ppup
    consensus = Map.filter (>= quorumN) votes
  ...
```

As an example: if a node "emitted" (however that happens) both `MaxBBSize 100`
and `MaxBBSize 200` in the same epoch it looks like that would count as a
different proposal than either just `MaxBBSize 100` or `MaxBSize 200` without
the other.

And if enough nodes emitted both `MaxBBSize 100` and `MaxBBSize 200`, then it
looks like that pair of seemingly conflicting `Ppm`s would -- as a whole
`PParamsUpdate` -- meet quorum, and when applied to the parameters, it would be
equiavlent to `MaxBBSize 200` (due to semantics of `Set.foldr` and the derived `Ord Ppm`).

```
updatePParams :: PParams -> PParamsUpdate -> PParams
updatePParams ppms (PParamsUpdate up) = Set.foldr updatePParams' ppms up
 where
  updatePParams' (MinFeeA               p) pps = pps { _minfeeA = p }
  updatePParams' (MinFeeB               p) pps = pps { _minfeeB = p }
  updatePParams' (MaxBBSize             p) pps = pps { _maxBBSize = p }
  ...
```

*tl; dr*

In summary, that all seems like many details and corner-cases to try to explain in this document's paragraph. Also, this paragraph focuses on a node as the primary actor, and I couldn't yet determine what this part of the Shelley interface looks like from a lone node's perspective. E.G. Will there be a transaction type that emits a single `Ppm`? Or is it less granular than that?

P.S. - Maybe it's too late or maybe it's already been considered, but perhaps replace `Set Ppm` by a more precise "dependent map"-ish type? It wouldn't be too bad given that the parameters and their types are all statically known.

Reply:
@edsko, FYI: @JaredCorduan and I had a call, and he clarified things for me. As a result, I wrote commit 2f9b7bbe42ccdd71577a03f1bc7a2a47601aff49 which you can cherry-pick/squash onto this PR if you'd like.

It rewrites this bullet list in the `HardFork.md`.

```
* 2f9b7bbe - (HEAD -> nfrisby/amend-hardforkdocs, origin/nfrisby/amend-hardforkdocs) HardFork.md: amend summary of Shelley protocol parameter update proposals (4 minutes ago) <Nicolas Frisby>
```

Reply:
looks great @nfrisby , thanks!

it would indeed be great to replace the `Set Ppm` by a "dependent map"-ish type.

Reply:
Ok, have cherry-picked that commit. Thanks :)